include/ddc/for_each.hpp

```diff
@@ -22,13 +22,13 @@ namespace ddc {
 namespace detail {
 
 template <class RetType, class Element, std::size_t N, class Functor, class... Is>
-void for_each_serial(
+KOKKOS_FUNCTION void for_each_serial(
```

Comment:
don't you end up with the previous problem if the provided functor is CPU-only even if it is called from CPU only?

Reply:
Good question, actually I did not get any warning for now.

I forced the compiler to compile a host-only `operator()` and indeed I got a warning.

Reply:
I see multiple options:
1. we keep this implementation and hope it does not trigger too many warnings, a bit disappointing though
2. we keep this implementation and ask people to always use `KOKKOS_LAMBDA`, does not cover our usage in gysela
3. we ask people to use it only inside a parallel algorithm, does not cover our usage in gysela
4. we implement `host_for_each` that should be used when it is called inside a host-only function (note that we should have the same problem when called inside a device-only function, should be quite rare though)
5. we implement `host_for_each` and a `parallel_for_each` that should accept a `ThreadHandle` that you get only from a `parallel_*` algorithm.

I guess Kokkos algorithms don't have this issue because they don't have sequential algorithms.

Reply:
Some piece of information https://llvm.org/docs/CompileCudaWithLLVM.html#overloading-based-on-host-and-device-attributes about what they call the "wrong-side rule"

Reply:
What about having a different implementation for host and for device? something like:
```c++
#ifdef ON_HOST
using current_execution_space = Kokkos::DefaultExecutionSpace;
#else
using current_execution_space = Kokkos::DefaultHostExecutionSpace;
#endif

template <class ExecutionSpace = current_execution_space, class RetType, class Element, std::size_t N, class Functor, class... Is>
struct for_each;

template <class RetType, class Element, std::size_t N, class Functor, class... Is>
struct for_each <Kokkos::DefaultExecutionSpace, RetType,Element,N, Functor, Is...> {
KOKKOS_FUNCTION void operator()(...);
};

template <class RetType, class Element, std::size_t N, class Functor, class... Is>
struct for_each <Kokkos::DefaultHostExecutionSpace, RetType,Element,N, Functor, Is...> {
void operator()(...);
};
```

Reply:
How the user would call it ? With a global object like ranges ?

Reply:
I think I have implemented your proposal, see https://godbolt.org/z/r7cK5cfd8. It emits the same warning.

To make it work we would need to know whether the host or the device is calling the `for_each` function but I am not aware of any way to access this information with CUDA.

Reply:
Indeed, my understanding of `__host__` behavior for nvcc seems wrong, Cf. https://godbolt.org/z/xddsdKvGx
I'm afraid there's no good solution

Reply:
Don't know if it suits your problem but macros KOKKOS_IF_ON_DEVICE() and KOKKOS_IF_ON_HOST() allow to execute different codes depending of if a KOKKOS_FUNCTION is called from host or device.

Reply:
Unfortunately it is not enough. The problem is that on the device compilation pass, `nvcc` sees that we call a host function from a host-device fonction therefore emits a warning. I do not think we can do much about it. I will make a host-only `host_for_each` function.